### PR TITLE
Fixing the index segment size check when writing PersistentIndex

### LINKED ISSUE #501: Index segments are persisted in the wrong order
In `IndexPersistor` in `PersistentIndex.java`, the index segments that are not current are persisted in the wrong order i.e. latest to oldest. The order of persistence has to be oldest to latest.

### LINKED ISSUE #790: Fix bugs in IndexPersistor in PersistentIndex
The persistor has two bugs

1) It refers to the size of the index segments to begin the persistence process - at this point the size may be 1 so the persistence begins and the last entry is the only entry in the map. Once the preprocessing is done, the persistor once again refers to the size to determine if there are any "previous" segments to write. At this point if the size is  > 1, the code will fail because it tries to do a `lowerEntry()` with the entry it picked up when the size was 1. Therefore the size used at both places must be the same.
2) It persists the "previous" segments in the wrong order (latest first). If there is a crash b/w the persistence of a later segment and an older segment, recovery will not be able to recover those blobs into the index because it only looks at the end offset of the index. Therefore, it should persist earliest first.

### PR DESCRIPTION
**Pull request by vgkholla**
Fixes #790 and Fixes #501 

Also fixes a lot of duplicate put record handling in the `StoreCopier`, `DiskReformatter` and `ConsistencyCheckerTool`

### CODE REVIEW

**Comment by pnarayanan** [comment]
Could you describe what is getting fixed here?

**Comment by vgkholla** [comment]
@pnarayanan fixed merge conflicts - good  to review

**Comment by vgkholla** [comment]
@pnarayanan Reminder

File: ambry-store/src/test/java/com.github.ambry.store/IndexTest.java [hunk 1]
```diff
@@ -1117,6 +1117,12 @@ public void testIndexSegmentRollOverKeySizeAndValueSizeChange() throws StoreExce
    */
   @Test
   public void indexPersistorTest() throws InterruptedException, IOException, StoreException {
+    // make sure persistor is persisting all segments
+    int numIndexSegments = state.index.getIndexSegments().size();
+    state.index.persistIndex();
+    int numIndexFiles = tempDir.listFiles(PersistentIndex.INDEX_SEGMENT_FILE_FILTER).length;
+    assertEquals("Not as many files written as there were index segments", numIndexSegments, numIndexFiles);
```
Comment (pnarayanan) [inline ambry-store/src/test/java/com.github.ambry.store/IndexTest.java]: Is there a way to use file access time or something to check for the order of persisting?

File: ambry-tools/src/main/java/com.github.ambry/store/ConsistencyCheckerTool.java [hunk 2]
```diff
@@ -172,31 +174,54 @@ public ConsistencyCheckerTool(ClusterMap clusterMap, StoreKeyFactory storeKeyFac
    * @return {@code true} if no real inconsistent blobs. {@code false}
```
Comment (pnarayanan) [inline ambry-tools/src/main/java/com.github.ambry/store/ConsistencyCheckerTool.java]: `@return` needs fixing. Any changes needed to the description?

File: ambry-tools/src/main/java/com.github.ambry/store/ConsistencyCheckerTool.java [hunk 3]
```diff
@@ -172,31 +174,54 @@ public ConsistencyCheckerTool(ClusterMap clusterMap, StoreKeyFactory storeKeyFac
...
+   */
+  private Pair<Boolean, Map<File, DumpIndexTool.IndexProcessingResults>> getIndexProcessingResults(File[] replicas)
+      throws Exception {
+    long currentTimeMs = time.milliseconds();
+    Map<File, DumpIndexTool.IndexProcessingResults> results = new HashMap<>();
+    boolean sane = true;
+    for (File replica : replicas) {
+      logger.info("Processing segment files for replica {} ", replica);
+      DumpIndexTool.IndexProcessingResults result =
+          dumpIndexTool.processIndex(replica, filterSet, currentTimeMs, throttler);
```
Comment (pnarayanan) [inline ambry-tools/src/main/java/com.github.ambry/store/ConsistencyCheckerTool.java]: what does `processIndex()` involve? (no javadocs for that method).

File: ambry-tools/src/main/java/com.github.ambry/store/ConsistencyCheckerTool.java [hunk 4]
```diff
@@ -172,31 +174,54 @@ public ConsistencyCheckerTool(ClusterMap clusterMap, StoreKeyFactory storeKeyFac
...
+      sane = sane && result.isIndexSane();
+      results.put(replica, result);
```
Comment (pnarayanan) [inline ambry-tools/src/main/java/com.github.ambry/store/ConsistencyCheckerTool.java]: any value in continuing if a previous result is not sane?

File: ambry-store/src/test/java/com.github.ambry.store/IndexTest.java (hunk 1 above)
Comment (vgkholla) [reply ambry-store/src/test/java/com.github.ambry.store/IndexTest.java]: I added a test but I'm not sure how useful it is. Though java says that the `File#lastModified()` provides a value in ms, it is limited by the underlying OS. On my mac, all I get are values to the closest second and the whole test runs in half a second, so it is not much use.

I put the test in a codebase that has the old code and it passes there too

The only way to conclusively test this would have been to insert hooks into the `IndexSegment` objects to determine when persist is called but with the current code setup where `PersistentIndex` creates these objects internally, there is no way to mock them or insert hooks.

File: ambry-tools/src/main/java/com.github.ambry/store/ConsistencyCheckerTool.java (hunk 2 above)
Comment (vgkholla) [reply ambry-tools/src/main/java/com.github.ambry/store/ConsistencyCheckerTool.java]: Changed

File: ambry-tools/src/main/java/com.github.ambry/store/ConsistencyCheckerTool.java (hunk 3 above)
Comment (vgkholla) [reply ambry-tools/src/main/java/com.github.ambry/store/ConsistencyCheckerTool.java]: Unfortunately many of the functions in the older tools don't have docs. In a nutshell, process index goes through all index files and
1) Determines if there are any duplicate puts/deletes/out of order entries
2) The state of a key in the index (valid/deleted/expired)
3) If some key is in a "recent" index segment

File: ambry-tools/src/main/java/com.github.ambry/store/ConsistencyCheckerTool.java (hunk 4 above)
Comment (vgkholla) [reply ambry-tools/src/main/java/com.github.ambry/store/ConsistencyCheckerTool.java]: Yeah - just to get results for all of them in case the caller wants to look and see and act upon it if required. 

For example, the `DiskReformatter` skips sanity checks if the source itself is not sane. It sends both the src and the newly created store files as "replicas" - it needs results for both.

**Comment by vgkholla** [comment]
@pnarayanan comments addressed

File: ambry-tools/src/main/java/com.github.ambry/store/StoreCopier.java [hunk 5]
```diff
@@ -224,7 +225,7 @@ public void close() throws IOException {
       return;
     }
     try {
-      shutDownExecutorService(scheduler, 1, TimeUnit.SECONDS);
+      shutDownExecutorService(scheduler, 5, TimeUnit.MINUTES);
```
Comment (pnarayanan) [inline ambry-tools/src/main/java/com.github.ambry/store/StoreCopier.java]: what is the reason behind this change?

File: ambry-tools/src/main/java/com.github.ambry/store/StoreCopier.java [hunk 6]
```diff
@@ -237,12 +238,14 @@ public void close() throws IOException {
...
    *                   destination.
-   * @return the {@link FindToken} until which data has been copied.
+   * @return a {@link Pair} of the {@link FindToken} until which data has been copied and a {@link Boolean} indicating
+   * whether the source had problems that were skipped over - like duplicates ({@code true} indicates that there were).
    * @throws IOException if there is any I/O error while copying.
    * @throws StoreException if there is any exception dealing with the stores.
    */
-  public FindToken copy(FindToken startToken) throws IOException, StoreException {
-    FindToken lastToken = null;
+  public Pair<FindToken, Boolean> copy(FindToken startToken) throws IOException, StoreException {
```
Comment (pnarayanan) [inline ambry-tools/src/main/java/com.github.ambry/store/StoreCopier.java]: Is this method ever called with a `startToken` that is not a `newFindToken()`? If so, how is it expected to be used? Also, what does the comment for `startToken` that says "does not perform any duplication checks at destination" mean - since you do seem to make such a check? It seems to me that your earlier comment that says "until which data has been copied" made sense, because this method cannot accept *any* token (unlike replication which is idempotent) - because your `else` case at 277 will not be correct in that case.

File: ambry-tools/src/main/java/com.github.ambry/store/StoreCopier.java (hunk 5 above)
Comment (vgkholla) [reply ambry-tools/src/main/java/com.github.ambry/store/StoreCopier.java]: It is possible that at the time the `StoreCopier` is done, there are multiple index files waiting to be written to disk. It was observed (empirically) that sometimes 1s wasn't enough.

File: ambry-tools/src/main/java/com.github.ambry/store/StoreCopier.java (hunk 6 above)
Comment (vgkholla) [reply ambry-tools/src/main/java/com.github.ambry/store/StoreCopier.java]: You raise a fair point - before all these changes I had envisioned being able to restart from a checkpoint but I realize now that it is not very useful and is probably more trouble that it is worth. I'll change this to not take in a `startToken` and update all the documentation.

Comment (vgkholla) [reply ambry-tools/src/main/java/com.github.ambry/store/StoreCopier.java]: Actually, it is still useful for this to take a `startToken` but I will change the doucumentation.

**Comment by vgkholla** [comment]
@pnarayanan Comments addressed - PTAL
